File: agent/promise_reflection.py
"""
Promise Reflection — session 結束時用 LLM 判斷哪些 unfulfilled actions 被兌現

對齊 src/lib/promise-reflection.ts:reflectAndMarkFulfilled。
紅線：
  - env PROMISE_REFLECTION_ENABLED=false 可全關
  - 失敗不阻斷（caller 應 catch）
  - 只標 confidence >= 4（防 hallucination）
"""
import json
import logging
import os
from datetime import datetime, timezone, timedelta

import firebase_admin
from firebase_admin import firestore

logger = logging.getLogger(__name__)

REFLECTION_MODEL = "claude-haiku-4-5-20251001"
MIN_CONFIDENCE = 4


def _enabled() -> bool:
    return os.environ.get("PROMISE_REFLECTION_ENABLED", "true").lower() != "false"


def _label(action_type: str) -> str:
    return {
        "promise": "我答應過",
        "question": "我問過",
        "event": "他/她的事",
        "note": "記得",
    }.get(action_type, "")


def _build_prompt(transcript: str, actions: list[dict]) -> str:
    lines = []
    for i, a in enumerate(actions, 1):
        label = _label(a.get("actionType") or "")
        body = a.get("title") or a.get("content") or ""
        prefix = f"（{label}）" if label else ""
        lines.append(f"{i}. [id={a['id']}] {prefix}{body}")
    return (
        "以下是一段對話記錄。你是一個誠實的紀錄員。\n\n"
        "【未兌現的承諾/問題/記得清單】\n"
        + "\n".join(lines)
        + "\n\n【對話記錄】\n"
        + transcript
        + "\n\n"
        "請判斷對話中是否兌現了清單中的條目（也就是角色實際聊了那個主題、回應了那個問題、提到了那個記得的事）。\n\n"
        "回 JSON 陣列，**每條清單項目都要評估**（不能漏、不能多）：\n"
        '[{"actionId":"<id>","fulfilled":true|false,"confidence":1-5}]\n\n'
        "confidence 標準：\n"
        "- 5 = 確定（對話明確處理了該條）\n"
        "- 4 = 高度可能\n"
        "- 3 = 模糊\n"
        "- 2 = 不太像\n"
        "- 1 = 完全沒提\n\n"
        "只回 JSON 陣列，不要其他文字、不要 code fence。"
    )


def _parse_verdicts(raw: str) -> list[dict]:
    cleaned = raw.strip()
    if cleaned.startswith("```"):
        # 容忍 code fence
        cleaned = cleaned.split("```")[1] if "```" in cleaned[3:] else cleaned[3:]
        if cleaned.startswith("json"):
            cleaned = cleaned[4:]
        cleaned = cleaned.strip().rstrip("```").strip()
    try:
        arr = json.loads(cleaned)
        if not isinstance(arr, list):
            return []
        return [
            v for v in arr
            if isinstance(v, dict)
            and isinstance(v.get("actionId"), str)
            and isinstance(v.get("fulfilled"), bool)
            and isinstance(v.get("confidence"), (int, float))
        ]
    except Exception:
        return []


def _ensure_init():
    if not firebase_admin._apps:
        from agent.firestore_loader import _ensure_init as _init
        _init()


def _load_unfulfilled_actions(
    character_id: str,
    user_id: str,
    days: int = 7,
    limit: int = 20,
) -> list[dict]:
    """撈該 (角色, 用戶) 的 unfulfilled + relevant actions

    對齊 character-actions.ts:getRecentUserActions（不帶 isRelevant 視為 True）
    """
    _ensure_init()
    db = firestore.client()
    q = (
        db.collection("platform_insights")
        .where("characterId", "==", character_id)
        .where("userId", "==", user_id)
        .where("fulfilled", "==", False)
        .limit(max(limit * 3, 30))
    )
    cutoff = datetime.now(timezone.utc) - timedelta(days=days)
    items = []
    for d in q.get():
        data = d.to_dict() or {}
        if data.get("isRelevant") is False:
            continue
        created = data.get("createdAt") or ""
        if not created:
            continue
        try:
            created_dt = datetime.fromisoformat(str(created).replace("Z", "+00:00"))
            if created_dt < cutoff:
                continue
        except Exception:
            continue
        items.append({
            "id": d.id,
            "actionType": data.get("actionType") or "general",
            "title": str(data.get("title") or ""),
            "content": str(data.get("content") or ""),
            "createdAt": created,
        })
    items.sort(key=lambda a: a.get("createdAt") or "", reverse=True)
    return items[:limit]


def reflect_and_mark_fulfilled(
    character_id: str,
    user_id: str,
    transcript: str,
    anthropic_api_key: str,
) -> dict:
    """執行 reflection 並回傳統計

    Returns: {enabled, checked, marked, skipped, errors}
    """
    stats = {"enabled": _enabled(), "checked": 0, "marked": 0, "skipped": 0, "errors": 0}
    if not stats["enabled"]:
        return stats
    if not transcript or len(transcript) < 50:
        return stats
    if not anthropic_api_key:
        return stats

    actions = _load_unfulfilled_actions(character_id, user_id)
    stats["checked"] = len(actions)
    if not actions:
        return stats

    try:
        import anthropic as anthropic_sdk
        client = anthropic_sdk.Anthropic(api_key=anthropic_api_key)
        resp = client.messages.create(
            model=REFLECTION_MODEL,
            max_tokens=600,
            messages=[{"role": "user", "content": _build_prompt(transcript, actions)}],
        )
        raw = resp.content[0].text or ""
        verdicts = _parse_verdicts(raw)
    except Exception as e:
        stats["errors"] += 1
        logger.warning(f"promise-reflection LLM failed: {e}")
        return stats

    valid_ids = {a["id"] for a in actions}
    db = firestore.client()
    for v in verdicts:
        action_id = v.get("actionId")
        if action_id not in valid_ids:
            stats["skipped"] += 1
            continue
        if not v.get("fulfilled"):
            stats["skipped"] += 1
            continue
        if v.get("confidence", 0) < MIN_CONFIDENCE:
            stats["skipped"] += 1
            continue
        try:
            db.collection("platform_insights").document(action_id).update({
                "fulfilled": True,
                "fulfilledAt": datetime.now(timezone.utc).isoformat(),
                "fulfilledBy": "auto-haiku",
            })
            stats["marked"] += 1
        except Exception as e:
            stats["errors"] += 1
            logger.warning(f"markActionFulfilled {action_id} failed: {e}")
    return stats

File: src/lib/promise-reflection.ts
/**
 * Promise Reflection — session 結束時用 LLM 判斷哪些 unfulfilled actions 被兌現
 *
 * 江彬實證：自動標 fulfilled 比手動可靠（手動沒人標）。LLM 看完整 transcript +
 * unfulfilled 清單，比角色自己 tool call 更不會漏。
 *
 * 防 hallucination：只標 confidence >= 4。
 *
 * 紅線：
 *   - env PROMISE_REFLECTION_ENABLED=false 可全關
 *   - 失敗不阻斷（caller 應 catch）
 *   - 已 fulfilled 的不重複跑
 */
import Anthropic from '@anthropic-ai/sdk';
import { getAnthropicClient } from './anthropic-via-bridge';
import {
  getRecentUserActions,
  markActionFulfilled,
  type CharacterAction,
} from '@/lib/character-actions';

const REFLECTION_MODEL = 'claude-haiku-4-5-20251001';
const MIN_CONFIDENCE = 4;

export interface ReflectionStats {
  enabled: boolean;
  checked: number;
  marked: number;
  skipped: number;
  errors: number;
}

interface LLMVerdict {
  actionId: string;
  fulfilled: boolean;
  confidence: number;
}

function isEnabled(): boolean {
  return process.env.PROMISE_REFLECTION_ENABLED !== 'false';
}

function buildPrompt(transcript: string, actions: CharacterAction[]): string {
  const lines = actions.map((a, i) => {
    const label = a.actionType === 'promise' ? '我答應過'
      : a.actionType === 'question' ? '我問過'
      : a.actionType === 'event' ? '他/她的事'
      : a.actionType === 'note' ? '記得'
      : '';
    const body = a.title || a.content || '';
    return `${i + 1}. [id=${a.id}] ${label ? `（${label}）` : ''}${body}`;
  });

  return `以下是一段對話記錄。你是一個誠實的紀錄員。

【未兌現的承諾/問題/記得清單】
${lines.join('\n')}

【對話記錄】
${transcript}

請判斷對話中是否兌現了清單中的條目（也就是角色實際聊了那個主題、回應了那個問題、提到了那個記得的事）。

回 JSON 陣列，**每條清單項目都要評估**（不能漏、不能多）：
[{"actionId":"<id>","fulfilled":true|false,"confidence":1-5}]

confidence 標準：
- 5 = 確定（對話明確處理了該條）
- 4 = 高度可能
- 3 = 模糊
- 2 = 不太像
- 1 = 完全沒提

只回 JSON 陣列，不要其他文字、不要 code fence。`;
}

function parseVerdicts(raw: string): LLMVerdict[] {
  let cleaned = raw.trim();
  if (cleaned.startsWith('```')) {
    cleaned = cleaned.replace(/^```[a-z]*\n?/i, '').replace(/```\s*$/, '').trim();
  }
  try {
    const arr = JSON.parse(cleaned);
    if (!Array.isArray(arr)) return [];
    return arr.filter(
      v => v && typeof v.actionId === 'string'
        && typeof v.fulfilled === 'boolean'
        && typeof v.confidence === 'number'
    );
  } catch {
    return [];
  }
}

export async function reflectAndMarkFulfilled(opts: {
  characterId: string;
  userId: string;
  transcript: string;
  anthropicApiKey: string;
}): Promise<ReflectionStats> {
  const stats: ReflectionStats = {
    enabled: isEnabled(),
    checked: 0,
    marked: 0,
    skipped: 0,
    errors: 0,
  };

  if (!stats.enabled) return stats;
  if (!opts.transcript || opts.transcript.length < 50) return stats;
  if (!opts.anthropicApiKey) return stats;

  const actions = await getRecentUserActions(opts.characterId, opts.userId, {
    limit: 20,
    unfulfilledOnly: true,
  });
  stats.checked = actions.length;
  if (actions.length === 0) return stats;

  const client = getAnthropicClient(opts.anthropicApiKey);
  let verdicts: LLMVerdict[] = [];
  try {
    const resp = await client.messages.create({
      model: REFLECTION_MODEL,
      max_tokens: 600,
      messages: [{ role: 'user', content: buildPrompt(opts.transcript, actions) }],
    });
    const raw = (resp.content[0] as Anthropic.TextBlock).text || '';
    verdicts = parseVerdicts(raw);
  } catch (e) {
    stats.errors += 1;
    console.warn('promise-reflection LLM failed:', e);
    return stats;
  }

  const validIds = new Set(actions.map(a => a.id).filter(Boolean));
  for (const v of verdicts) {
    if (!validIds.has(v.actionId)) { stats.skipped += 1; continue; }
    if (!v.fulfilled) { stats.skipped += 1; continue; }
    if (v.confidence < MIN_CONFIDENCE) { stats.skipped += 1; continue; }
    try {
      await markActionFulfilled(v.actionId, 'auto-haiku');
      stats.marked += 1;
    } catch (e) {
      stats.errors += 1;
      console.warn(`markActionFulfilled ${v.actionId} failed:`, e);
    }
  }
  return stats;
}
